merkle/src/merkle.rs

```diff
@@ -248,6 +253,253 @@ impl<E: Element> Clone for MmapStore<E> {
+        // created there has an assigned size of *only* the leaves (the data,
+        // as it's seen there), not the *entire* tree, which is expected since
+        // the consumer shouldn't need to know how big the final `Store` is
+        // going to be. To *temporarily* accommodate this we *mangle* the
+        // received `size` (in a similar way to `MerkleTree::from_iter`) to
+        // expand it to the entire tree.
+        let pow = next_pow2(size);
+        let size = 2 * pow - 1;
+
+        let byte_len = E::byte_len() * size;
+        let file: File = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(&path)
+            .unwrap();
+        file.set_len(byte_len as u64).unwrap();
+
+        let mmap = unsafe { MmapMut::map_mut(&file).expect("couldn't create map_mut") };
+        let mmap_size = mmap.len();
+        DiskMmapStore {
+            store: Arc::new(RwLock::new(Some(mmap))),
+            len: 0,
+            _e: Default::default(),
+            file,
+            store_size: mmap_size,
+            path: path.to_str().expect("couldn't find path").to_string(),
+            size: Some(size),
+        }
+    }
+
+    pub fn store_size(&self) -> usize {
+        return self.store_size;
+    }
+
+    pub fn store_read_range(&self, start: usize, end: usize) -> Vec<u8> {
+        self.reload_store();
+        // FIXME: Not actually thread safe, the `store` could have been offloaded
+        //  after this call (but we're not striving for thread-safety at the moment).
+
+        match *self.store.read().unwrap() {
+            Some(ref mmap) => return mmap[start..end].to_vec(),
+            None => panic!("The store has not been reloaded"),
+        }
+        // FIXME: Remove clone, this is duplicating lots of data.
+    }
+
+    pub fn store_copy_from_slice(&self, start: usize, end: usize, slice: &[u8]) {
+        self.reload_store();
+        match *self.store.write().unwrap() {
+            Some(ref mut mmap) => return mmap[start..end].copy_from_slice(slice),
+            None => panic!("The store has not been reloaded"),
+        }
+    }
+
+    // Checks if the `store` is loaded and reloads it if necessary.
+    // FIXME: Check how to compact this logic.
+    fn reload_store(&self) {
+        let mut need_to_reload_store = false;
+        {
+            if self.store.read().unwrap().is_none() {
+                need_to_reload_store = true;
+            }
+        }
+        if need_to_reload_store {
+            let mut mmap = self.store.write().unwrap();
+
+            let new_store: DiskMmapStore<E> = DiskMmapStore::new_with_path(self.size.expect("couldn't find size"), Path::new(&self.path));
+//            self.store = Arc::new(RwLock::new();
+            // FIXME: Extract part of the `MmapMut` creation logic to avoid
+            //  recreating the entire `DiskMmapStore`.
+
+            // FIXME: How to store the new mmap structure? Both options are
```

Comment:
@dignifiedquire Could you also take a look at this please?

Reply:
I pushed something that makes it compile

Reply:
Unblocked (twice in a day), and learned a new Rust trick, thanks!